sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,750 @@
+to an individual output module, and each element of a row is a name of a file
+containing necessary information about the module, such as the code itself, its
+properties.
+
+At the action graph building stage for each requested AOT compilation target -
+SPIR-V-based (such as Gen targets) and/or non-SPIR-V-based (such as PTX) - the
+driver adds an `aspect-filter` action which filters out input file table rows
+with device code modules using features unsupported on current target. Then the
+output table goes as input into the AOT stage, and the prior filtering
+guarantees that the AOT compiler will not encounter device code it can't
+compile. In the extreme case when all device code
+modules use unsupported aspects, the input file table will be empty. The picture
+below illustrates the action graph built by the clang driver along with file
+lists and tables generated and consumed by various nodes of the graph. The
+example set of targets used for the illustration is 4 targets
+- spir64 (runtime JITted SPIR-V)
+- AOT targets
+    - non-SPIR-V based
+        - nvptx (PTX)
+    - SPIR-V based
+        - gen_11 (Intel Gen)
+        - x86_64_avx512 (AVX512)
+<br>
+<br>
+
+![Device SPIRV translation and AOT compilation](images/DeviceLinkAOTAndWrap.svg)
+
+##### Aspect filter tool
+
+This tool transforms an input file table by removing rows with device code files
+that use features unsupported for the target architecture given as tool's
+argument.
+
+*Input*:
+- file table, normally coming out of `sycl-post-link` or `file-table-tform`
+  tools
+
+*Command line arguments*
+- target device architecture to filter for
+- path to the device configuration file
+
+*Output*
+- the input file table filtered as needed
+
+In more details, the tool performs the following actions:
+1) Checks if the input file table contains "Properties" column. If not, copies
+   the input file table to output and exits without error.
+1) Reads in the device configuration file and finds some entry `E` corresponding
+   to the architecture given on the command line. If there is no such entry -
+   reports and error and exits.
+1) For each row in the input file table:
+   - loads the properties file from the "Properties" column
+   - checks if there is the `SYCL/image-requirements` property
+   - if no, copies current row to the output file table and goes to the next
+   - if yes, checks if all the requirements listed in the property are supported
+     by the target architecture as specified in the device configuration file
+       - if yes, copies current row to the output file table and goes to the
+         next
+       - otherwise skips this row
+
+##### Configuration file location and driver option
+
+A default device configuration file is supplied as a part of OneAPI SDK. It is
+located in the TBD directory. Users may override the defalt using the
+```
+-fsycl-device-config-file=<path>
+```
+option.
```

Comment:
There is an important part of design that is not described anywhere.  The parameters to the `-fsycl-targets` command line option need to change.  Currently, those parameters must be one of the following strings:

* spir64-unknown-unknown-sycldevice
* spir64_x86_64-unknown-unknown-sycldevice
* spir64_fpga-unknown-unknown-sycldevice
* spir64_gen-unknown-unknown-sycldevice

But, it seems like this design only works if we change the parameter to match the strings that are used as device names in the configuration file (`gen11_1`, `gen_icl`, etc.)

In addition, we currently tell users to pass an option like `-Xs -device *` in order to specify the exact device to generate AOT code for.  This won't work anymore, so I think the design doc should state this.

Reply:
> The parameters to the -fsycl-targets command line option need to change.

certainly. As I mentioned, I would very much prefer that all arch specifications are consistent across various places - aspect enums, config file, arch specification via -fsycl-targets.
To maintain backward compatibility, we need to still support the long older variants too.

Reply:
Created https://github.com/gmlueck/llvm/pull/1